### PR TITLE
When watching failed lookups, watch packageDir if its a symlink otherwise the path we use to watch

### LINKED ISSUE #55450: Having to restart TS server every time I build a package in a workspace
Type: <b>Bug</b>

I have a monorepo setup, using workspaces. 
Whenever I rebuild one of my packages, all files in another workspace can no longer resolve types for it. That is until I do cmd + shift + P -> TS restart

VS Code version: Code 1.81.0 (Universal) (6445d93c81ebe42c4cbd7a60712e0b17d9463e97, 2023-08-02T12:40:02.782Z)
OS version: Darwin arm64 21.4.0
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M1 Pro (10 x 24)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|7, 9, 9|
|Memory (System)|32.00GB (0.07GB free)|
|Process Argv|--crash-reporter-id 971cdb5c-5eac-4e31-85ae-d33323f572d4|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (9)</summary>

Extension|Author (truncated)|Version
---|---|---
postcss|css|1.0.9
vscode-eslint|dba|2.4.2
gitlens|eam|14.2.1
prettier-vscode|esb|10.1.0
vscode-jest-runner|fir|0.4.66
angular-console|nrw|18.6.0
vscode-stylelint|sty|1.2.4
vscode-mdx|uni|1.4.0
vscode-css-variables|vun|2.6.3

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscoreces:30445986
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
pythonvs932:30410667
py29gd2263:30792226
vsclangdf:30486550
c4g48928:30535728
dsvsc012:30540252
pynewext54:30695312
azure-dev_surveyone:30548225
vscccc:30803845
282f8724:30602487
89544117:30613380
showlangstatbar:30737416
vsctsb:30748421
03d35959:30757346
pythonfmttext:30731395
pythoncmvfstrcf:30756944
fixshowwlkth:30771522
showindicator:30805244
pythongtdpath:30769146
i26e3531:30792625
gsofa:30804715
pythonnosmt12:30797651
pythonidxpt:30805730
pythonnoceb:30805159
e537b577:30795824
dsvsc013:30795093
dsvsc014:30804076

```

</details>

### PR DESCRIPTION
**Pull request by sheetalkamat**
In mono repo scenarios the packages are linked using symlink. 
So when trying to resolve `package1` from `package2` in the test we look for locations like: `/home/src/projects/project/node_modules/package1/dist/index.d.ts` 

As a result when `package1` fails to resolve, we want to watch if this path will exists.
Before this PR, we would be watching `/home/src/projects/project/node_modules` as its a `node_modules` folder and we dont want to be watching too many packages or folders. This results in issues because `/home/src/projects/project/node_modules/package1` is a symlink and any changes to the target are not percolated to the watcher we create. So without this PR we would not see the changes and miss the package build and report errors about `package1` not found and never be able to resolve that unless offcourse if you start fresh by restarting tsserver

Now when we calculate dir to watch, we also find the `packageDir` if this happens to be symlink, we will watch it otherwise we will watch the `node_modules` like we use to watch before. 

Other change needed for this to work on linux was that when we are watching this symlink dir, we need to ensure we are watching the target recursively as that is the expectation for native watchers that support symlink watching. 

We already had test case which didnt update the diagnostics on tsserver when `paths` was not used and with my other prototype changes to treat file create and delete as change and retain things more, it becomes more evident and those tests just start to fail. So this is cruicial fix for that change to go in as well

Fixes #55450

### CODE REVIEW

File: tests/baselines/reference/tsserver/projectReferences/monorepo-like-with-symlinks-when-referencing-file-from-subFolder-with-scoped-package-and-solution-is-not-built-with-preserveSymlinks.js [hunk 1]
```diff
@@ -272,7 +272,7 @@ FsWatches::
   {}
 
 FsWatchesRecursive::
-/user/username/projects/myproject/node_modules: *new*
+/user/username/projects/myproject/node_modules/@issue/b: *new*
```
[the remark this answers is not in the document]
Comment (sheetalkamat) [reply tests/baselines/reference/tsserver/projectReferences/monorepo-like-with-symlinks-when-referencing-file-from-subFolder-with-scoped-package-and-solution-is-not-built-with-preserveSymlinks.js:275]: Casing is handled. Its not target we are watching but symlink so this is something we lookedup as part of failed lookup and we are watching that.

**Review by jakebailey — APPROVED** [review]

**Comment by openscript** [comment]
Thank you very much for your work on this! Highly appreciated 🎉

**Comment by rivaros** [comment]
@sheetalkamat could you update when this fix will be released in typescript?

**Comment by sheetalkamat** [comment]
This is already in 5.5 beta

**Comment by rivaros** [comment]
@sheetalkamat So this fix is already in prod? 
Currently using TS 5.5.3

Still having the issue with VSCode. One of my node_modules packages - `node_modules/@libraryname/package` is a symbolic link to another folder. If files are changed in that folder, should it detect changes?

**Comment by sheetalkamat** [comment]
This is in 5.5
If you are running into issues, pls file issue with logs and way to repro.

**Pull request merged**